multiqc/core/ai.py

```diff
@@ -0,0 +1,764 @@
+                    "(see https://docs.seqera.io/multiqc/reports/customisation#hiding-columns) to reduce the context. "
+                    "You can also open the HTML report in the browser, hide columns or samples dynamically, and request "
+                    "the AI summary dynamically, or copy the prompt into clipboard and use it with extrenal services."
+                )
+                return user_prompt + genstats_context, True
+        user_prompt += genstats_context
+        current_n_tokens += genstats_n_tokens
+
+    user_prompt = re.sub(r"\n\n\n", "\n\n", user_prompt)  # strip triple newlines
+
+    # Build sections context and use it unless it's too long for the LLM context
+    sec_context: str = ""
+    for section in metadata.sections.values():
+        tool = metadata.tools[section.module_anchor]
+        sec_context += "\n----------------------\n\n"
+        sec_context += f"Tool: {tool.name}\n"
+        sec_context += f"Section: {section.name}\n"
+        if section.description:
+            sec_context += f"Section description: {_strip_html(section.description)}\n"
+        if section.comment:
+            sec_context += f"Section comment: {_strip_html(section.comment)}\n"
+        if section.helptext:
+            sec_context += f"Section help text: {_strip_html(section.helptext)}\n"
+
+        if section.content_before_plot:
+            sec_context += section.content_before_plot + "\n\n"
+        if section.content:
+            sec_context += section.content + "\n\n"
+
+        if section.plot_anchor and section.plot_anchor in report.plot_by_id:
+            plot = report.plot_by_id[section.plot_anchor]
+            if plot_content := plot.format_for_ai_prompt(keep_hidden=True):
+                if plot.pconfig.title:
+                    sec_context += f"Title: {plot.pconfig.title}\n"
+                sec_context += "\n" + plot_content
+
+        # Check if adding this section would exceed the limit
+        # Using rough estimate of 4 chars per token
+        sec_n_tokens = client.n_tokens(sec_context)
+        if current_n_tokens + sec_n_tokens > max_tokens:
+            logger.warning(
+                f"Truncating prompt to only the general stats to fit within {client.title}'s context window ({client.max_tokens()} tokens). "
+                f"Tokens estimate: {current_n_tokens}, with sections: at least {current_n_tokens + sec_n_tokens}"
+            )
+            return user_prompt, False
+
+    user_prompt += sec_context
+    user_prompt = re.sub(r"\n\n\n", "\n\n", user_prompt)  # strip triple newlines
+    return user_prompt, False
+
+
+def _save_prompt_to_file(prompt: str):
+    # Save content to file for debugging
+    path = report.data_tmp_dir() / "multiqc_ai_prompt.txt"
+    system_prompt = PROMPT_FULL if config.ai_summary_full else PROMPT_SHORT
+    path.write_text(f"{system_prompt}\n\n----------------------\n\n{prompt}")
+    logger.debug(f"Saved AI prompt to {path}")
```

Comment:
I think we can ditch this, as the file isn't visible after the run (for me at least), so not sure how useful it is.
```suggestion
```

Reply:
It should be visible, are you sure?
<img width="1258" alt="Screenshot 2025-01-17 at 10 42 14" src="https://github.com/user-attachments/assets/e5fc9416-5171-42fa-83d4-6109923b482a" />

I used it personally to debug. We can keep it only in `--dev` mode, but I thought it might be useful generally.

Reply:
Didn't see this reply sorry but figured it out: the log message shows the path to the `tmp` directory, which is then moved to `./multiqc_data/`. I didn't realise that it was the same one and tried to look at the file path that was in the verbose log message, but it's not there after the run as the tmp directory is removed after it's moved to the launch dir.

I've pushed a commit that changes the log message to just say `Saved AI prompt to 'multiqc_data'` as then I wouldn't have made this error (and the tmp path is meaningless if it doesn't exist).